### ISSUE TITLE
IUserRoleStore interface - make role TRole rather than a string

### ISSUE THREAD

**Issue by flyfishMT** [aspnet/Identity#483]
The IUserRoleStore interface's methods all use a string "roleName" to define a role.  I want to customize the role so there is more to it (and the relationship between user and role) than just the rolename; having the role object instead would allow this and provide more flexibility - 

    public interface IUserRoleStore<TUser, TRole> : IUserStore<TUser> 
    where TUser: class
    where TRole: class
    {
    	Task AddToRoleAsync(TUser user, TRole role, CancellationToken cancellationToken);
    	Task RemoveFromRoleAsync(TUser user, TRole role, CancellationToken cancellationToken);
    	Task<IList<TRole>> GetRolesAsync(TUser user, CancellationToken cancellationToken);
    	Task<bool> IsInRoleAsync(TUser user, TRole role, CancellationToken cancellationToken);
    	Task<IList<TUser>> GetUsersInRoleAsync(TRole role, CancellationToken cancellationToken);
    }

Methods in UserManager would need to change too - 

    public virtual Task<IList<TRole>> GetRolesAsync(TKey userId);
    public virtual Task<IdentityResult> AddToRoleAsync(TKey userId, TRole role);

Basically, my relationship between user and role is many-to-many. Role.UserId, Role.Name is not a unique key...

**Comment by HaoK** [comment]
That's what the RoleManager object is for.  The IUserRole store is just responsible for storing the fact that the user is part of the Role.

**Comment by flyfishMT** [comment]
The User to Role relationship is what is different in my implementation. So IUserRoleStore is what doesn't work for me. An example would be there are different roles by "UserGroup" - and the UserGroup's can have roles with the same name, and a user can be in more than one UserGroup.  

This:

    public virtual async Task<IdentityResult> RemoveFromRoleAsync(TUser user, string role)

Doesn't work, because a *user* has *role* twice, in 2 different usergroups.  I'm just trying to use the stock UserManager, maybe I can't.

**Comment by HaoK** [comment]
Yeah identity assumes username and role names are keys just like id. Maybe you could just Concat group + name to make it the key in this case?.

**Comment by flyfishMT** [comment]
ok thanks
